sympy/matrices/matrices.py

```diff
@@ -2402,18 +2400,24 @@ def dot(self, b):
                     type(b))
 
         mat = self
-        if mat.cols == b.rows:
-            if b.cols != 1:
-                mat = mat.T
-                b = b.T
-            prod = flatten((mat * b).tolist())
-            if len(prod) == 1:
-                return prod[0]
-            return prod
-        if mat.cols == b.cols:
-            return mat.dot(b.T)
-        elif mat.rows == b.rows:
-            return mat.T.dot(b)
+        m = len(mat)
+        n = len(b)
+        if m == n:
+            if (mat.rows == 1 or mat.cols == 1) and (b.rows == 1 or b.cols == 1):
+                if mat.cols == b.rows:
+                    if b.cols != 1:
+                        mat = mat.T
+                        b = b.T
+                    prod = flatten((mat * b).tolist())
+                    if len(prod) == 1:
+                        return prod[0]
+                    return prod
+                if mat.cols == b.cols:
```

Comment:
Once we know the sizes of the matrix are the same, `.reshape` can be used to force everything into the proper dimensions.